### DISCUSSION TITLE
Fuzzy Display

### DISCUSSION THREAD

**Discussion by stevehol** [ltsp/ltsp#301] [Q&A]
Most of our clients give a nice sharp display most of the time. Rarely they get a very fuzzy display right from the login display which continues to the desktop. 
But one machine gets this fuzzy display most of the time. Its an HP t620  - the only one we've got with Radeon 8280E graphics. We're booting an Xubuntu 20.04 chrootless image. About 25% of boots are good, 75% fuzzy.
I would suspect the hardware, except that everything I boot from a USB stick, including Xubuntu 20.04 is always OK.
When the display is fuzzy, `$ inxi -G` shows all the expected info, including Device: AMD/ATI Kabini [Radeon 8280E] driver: radeon resolution: 1920x1080~60Hz.
Its the same whether we use IMAGE_TO_RAM or not, and with different RAM modules.

**Comment by alkisg** [comment]
Hello, please upload your /var/log/Xorg.?.log from a live USB and from the same client booted via LTSP, so that we compare what's different there. Most probably you're using a different kernel/xorg, as LTSP by itself doesn't affect the display.

**Comment by stevehol** [comment]
@alkisg thanks for your help.
There are only slight differences in the kernel and xorg-server versions. The Xorg logs are at:  Live USB: https://pastebin.com/2eFKaxC4  LTSP: https://pastebin.com/6hpDuVER
The one big difference I notice is that the LTSP log has twice as many 'DDC gathered Modelines' tables - although the modeline values in all tables in both logs are identical.

**Comment by alkisg** [comment]
The kernel and xorg are a different version as well, 5.4.0-26-generic vs 5.4.0-40-generic and 2:1.20.8-2ubuntu2 vs 2:1.20.8-2ubuntu2.1.

Get a root shell in the client, and try this:

```shell
systemctl stop lightdm
xinit
```

This will just run an xterm, is this fuzzy?
Also, do upload a photo of the "fuzzy display", so that we see what we're talking about...

**Comment by stevehol** [comment]
![Screenshot](https://user-images.githubusercontent.com/64087867/87301457-83820280-c507-11ea-9dfb-2b67d7736351.jpg)
![2](https://user-images.githubusercontent.com/64087867/87301516-9e547700-c507-11ea-940d-027c29be30f9.jpg)
![3](https://user-images.githubusercontent.com/64087867/87301526-a6141b80-c507-11ea-9765-e0db8bfb74fb.jpg)
![4](https://user-images.githubusercontent.com/64087867/87301543-ae6c5680-c507-11ea-9ed1-31124d6488ed.jpg)
![5](https://user-images.githubusercontent.com/64087867/87301561-b6c49180-c507-11ea-9fee-8e7bbd47bcb8.jpg)

Going to the virtual console Ctrl-Alt-F2:-

```
$ sudo -i
# systemctl stop lightdm
# xinit

```
The xterm is fuzzy here too.

But I've noticed that if I leave it until the screensaver blanks the screen, then move the mouse to get it back - its suddenly sharp!

**Comment by stevehol** [comment]
The screenshot is perfectly sharp, but the display shown in the camera shots is not.

**Comment by alkisg** [comment]
If you press Ctrl+Alt+F2 and Ctrl+Alt+F7, does that restore the display?
Another thing to try would be to add "quiet splash" under [server] in ltsp.conf:

```shell
[server]
KERNEL_PARAMETERS="quiet splash"
```

Then run `ltsp ipxe` and see if the splash screen helps workaround the issue.
It's certainly not related to LTSP though, as LTSP doesn't try to configure Xorg by default...

**Comment by stevehol** [comment]
OK I will try those. If not this machine will just have to sit on the shelf until something gets fixed.
Thanks for your help.

**Comment by enaut** [comment]
This looks like a wrong connection between computer and display. Either replug/change your cables or in case of a laptop your display might be defect.

**Comment by stevehol** [comment]
@ enaut thanks, but it is definitely not a physical connection issue - it happens with all our clients, with different radeon graphics eg:  8330E, 8280E, R5E, and it is the same on DisplayPort and DVI.

**Comment by stevehol** [comment]
### Update
Since updating to 5.4.0-42-generic with xserver-xorg-core 2:1.20.8-2ubuntu2.2 ALL our clients are affected every boot.
However, installing Xubuntu 20.04 directly to SSD on a client, and updating to the same versions, does not show the problem.

### Workaround
I noticed that if I leave the fuzzy LightDM login screen to timeout (10 mins default) then when the mouse is moved to redisplay, it is sharp, and logging in to Xfce all remains sharp.
The timeout appears to be the X Screensaver - the only config I found to change it was to put 
```
Section "ServerFlags"
	Option "BlankTime" "1"
EndSection
```
into an xorg.conf file eg: /etc/X11/xorg.conf.d/50-BlankTime.conf
as suggested here: https://shallowsky.com/linux/x-screen-blanking.html
This workaround reduces the timeout to one minute so that users only have to wait 1 min for the screen to blank before they can login. 
(I now have to turn off DPMS with `xset -dpms` in each users ~/.xsessionrc to prevent a 1 min timeout by the Xfce screensaver).

**Comment by alkisg** [comment]
Xubuntu 20.04.1 just came out, it's a good chance to compare live USB boots vs LTSP boots as now they'll have the same Xorg/kernel versions.

Btw, do Intel clients get this issue, or does it only affect AMD ones?

**Comment by stevehol** [comment]
I'll try to test 20.04.1 and update.
We only have AMD clients - I don't have an Intel machine to test now!

**Comment by stevehol** [comment]
@alkisg 
Just tested a fresh install of Xubuntu 20.04.1 and LTSP on a Dell Wyse 5060 with AMD GX-424CC + Radeon R5E.

The server machine displays sharp (as they always have) but my AMD clients get the fuzzy display still :(

But an Intel client is OK - sharp display right from LightDM. (Intel Pentium N3710 with Intel HD 405 graphics).
Unfortunately all my working clients are AMD so I'm stuck with the login delay workaround.

**Comment by alkisg** [comment]
@stevehol, and does the problem happen with an updated local installation, or with a Xubuntu 20.04.1 live USB stick?

The idea was to compare local 20.04.1 vs netbooted 20.04.1...

Anyway this problem isn't caused by LTSP as LTSP doesn't configure Xorg by default, so you should be asking in some Xorg/kernel/amdgpu mailing list about it...

**Comment by stevehol** [comment]
I have only ever seen this issue on machines booted as LTSP clients.
The same machines booted from SSD installation, or live USB sticks, do not have this issue.
I have never seen this issue on any machine booted from an SSD installation or live USB stick.

The Dell Wyse 5060 on which I installed Xubuntu 20.04.1 and LTSP server, always has a sharp display booting from SSD. But this same machine, when booted as an LTSP client, always has the fuzzy / blurry display issue.

**Comment by alkisg** [comment]
For example, [gnome-keyring fails on SSHFS](https://bugzilla.gnome.org/show_bug.cgi?id=730587).

Since many people use SSHFS with LTSP but not with normal installations, they're thinking it's an LTSP issue.
But it's not, it cannot be solved in the LTSP code. It needs to be solved in the gkm-transaction.c file of the gnome-keyring code, which LTSP developers have no access to.

So even if you only see this in LTSP clients, it cannot be resolved in the LTSP code base because there's no code in LTSP that configures Xorg by default (well, unless you're using a custom xorg.conf on the server and you send it to the clients via the image).

You need to report it to the Xorg/AMD/kernel developers.

**Comment by stevehol** [comment]
> there's no code in LTSP that configures Xorg by default

 - but is it possible that the way LTSP boots clients, prevents or interferes with the usual initial hardware detection? (which normally works correctly for non-LTSP boots of the same hardware) Then after resuming from the X screensaver blanking 
perhaps X redetects the hardware and it is now effective? 

I have tried to find a combination of KERNEL_PARAMETERS and X_MODES etc which gets a sharp display of LightDM, but have been unsuccessful.

**Comment by alkisg** [comment]
> but is it possible that the way LTSP boots clients, prevents or interferes with the usual initial hardware detection?

There's no hardware-related code in LTSP. That said, in NFS root some latencies or file semantics might be slightly different than ext4 and could trigger a bug in xorg or amdgpu.

It's possible to boot a single client from an NFS read-write root with nfsroot=server:/path/to/chroot, without involving LTSP at all. You may find such tutorials online, it's rather easy, you can just unsquashfs your existing x86_64.img image netboot it. If you see that it happens with nfsrw root without LTSP, then you'll have proof that LTSP isn't involved at all.
